### ISSUE TITLE
"incomingtx" `event` is not triggered 

### ISSUE THREAD

**Issue by n8tb1t** [turtlecoin/turtlecoin-wallet-backend-js#70]
If I send coins from one subwallet to another in the same container - the "incomingtx" event is not triggered, only the `outgoingtx` is, and I see all the outgoing and incoming transactions in it.

**Comment by zpalmtree** [comment]
Hmm, nice find. Currently we just dispatch the events based on the total value of the transaction here: https://github.com/turtlecoin/turtlecoin-wallet-backend-js/blob/134c7a1e29137690a1983d01debf356a38bfe79b/lib/WalletBackend.ts#L2143-L2148

I think we probably can instead check the total value of each transfer to the subwallets to determine if we need to send multiple events.

However, it may be kind of confusing for developers if they get the same event multiple times for a single transaction, for example, if two subwallets receive funds in one transaction, and one spends funds, should we dispatch two `incomingtx`, and one `outgoingtx`? Or just one `incomingtx`, and one `outgoingtx`? If we did the former, it might make sense to add a `address` parameter to show which address the event relates to. 

What do you think?

**Comment by n8tb1t** [comment]
Also, `on balance update` trigger will be really useful, to update locked/unlocked balance in the DB. Now I have to run a scheduled check on wallets with locked balances.

**Comment by n8tb1t** [comment]
if two subwallets receive funds in one transaction, and one spends funds, we dispatch two incomingtx, and one outgoingtx. in my case the raw data is going straight into the db, with addresses already resolved, so I can query it later and find sender's and receiver's wallets and transaction by hash.  I guess in my case the simplest solution would be to use `on transaction` event and treat transaction with positive amount as incoming and with negative as outgoing.

**Comment by zpalmtree** [comment]
>Also, on balance update trigger will be really useful, to update locked/unlocked balance in the DB. Now I have to run a scheduled check on wallets with locked balances.

I'm a little wary adding that, assuming you're referring to locked balances due to things like coinbase transaction unlock times, and transaction unlock times, rather than just locked balance due to unconfirmed transactions.

We only calculate unlocked balance when the user requests it, and there is no wallet process which lets us figure out when a transaction unlocks naturally.

It would potentially be possible to store the blocks that transactions will unlock at, so we know when to emit the event internally, rather than having to poll every transaction on every block, which I imagine would slow down runtime an amount with a large amount of transactions. However, that doesn't take into account transactions that have a time based unlock, rather than a block based unlock.

Basically I'm happy to add it, if it can be done efficiently, which I'm not sure of.

**Comment by zpalmtree** [comment]
>if two subwallets receive funds in one transaction, and one spends funds, we dispatch two incomingtx, and one outgoingtx

OK. I'll update at some point to implement this behaviour.

**Comment by n8tb1t** [comment]
Thanks!:) Is there a formula to calculate the amount that will be locked, how it's calculated?

**Comment by zpalmtree** [comment]
Are you referring to the amount of balance being locked when you send a transaction?

**Comment by n8tb1t** [comment]
yes:)

**Comment by zpalmtree** [comment]
It's semi random. Your wallet is comprised of multiple differently valued "inputs", and we randomly select inputs of different values until we have enough to cover your transaction, and usually a bit over.

For example, if you want to send a transaction of 1234 TRTL, we might select an input of 20 TRTL, 1000 TRTL, and 500 TRTL. We would then send 1520 TRTL, with 1234 TRTL going to the destination, 0.1 TRTL going to fees, and 285.90 TRTL returning as change. 

This 285.90 TRTL is the amount which will be locked, as we need to wait for the transaction to be included in a block until we can spend those funds.

If you're only interested in locked balance updates due to this process, this *should* be easily handled by the `transaction` event, since once we download the block the sent transaction was in and process it, our balance will unlock and the corresponding events should be sent.

**Comment by n8tb1t** [comment]
so you say, the moment I see an incoming tx to the receiving wallet, the sender locked balance is already unlocked, and I can just update it?

**Comment by zpalmtree** [comment]
If it's the same transaction that you sent, then yes. In TonChan, I fire balance updates upon receiving a transaction, and it seems to work fine.
